include/geos/algorithm/LineIntersector.h

```diff
@@ -589,8 +589,20 @@ class GEOS_DLL LineIntersector {
     {
         geom::CoordinateXYZM ptInt(Intersection::intersection(p1, p2, q1, q2));
         if (ptInt.isNull()) {
-            // FIXME need to cast to correct type in mixed-dimensionality case
-            ptInt = static_cast<const C1&>(nearestEndpoint(p1, p2, q1, q2));
+            const geom::CoordinateXY& nearest = nearestEndpoint(p1, p2, q1, q2);
+#if __cplusplus >= 201703L
+            if constexpr (std::is_same<C1, C2>::value) {
+#else
+            if (std::is_same<C1, C2>::value) {
+#endif
+                ptInt = static_cast<const C1&>(nearest);
+            } else {
+                if (&ptInt == static_cast<const geom::CoordinateXY*>(&p1) || &ptInt == static_cast<const geom::CoordinateXY*>(&p2)) {
```

Comment:
Probably due to my lack of familiarity with that code, but I can't make sense of those address comparisons.  ptInt is a new object created at line 590. How could its address be the same as p1 or p2 ?

Reply:
It couldn't! The wrong variable is being compared here.